tests/cross-compile/slang-cpp-prelude.h

```diff
@@ -80,12 +504,187 @@ struct Vector<T, 4>
     T x, y, z, w;
 };
 
+
+typedef Vector<float, 2> float2;
+typedef Vector<float, 3> float3;
+typedef Vector<float, 4> float4;
+
+typedef Vector<int32_t, 2> int2;
+typedef Vector<int32_t, 3> int3;
+typedef Vector<int32_t, 4> int4;
+
+typedef Vector<uint32_t, 2> uint2;
+typedef Vector<uint32_t, 3> uint3;
+typedef Vector<uint32_t, 4> uint4;
+
 template <typename T, int ROWS, int COLS>
 struct Matrix
 {
     Vector<T, COLS> rows[ROWS];
 };
 
+// ----------------------------- ResourceType -----------------------------------------
+
+// https://docs.microsoft.com/en-us/windows/win32/direct3dhlsl/sm5-object-structuredbuffer-getdimensions
+// Missing  Load(_In_  int  Location, _Out_ uint Status);
+
+template <typename T>
+struct RWStructuredBuffer
+{
+    T& operator[](size_t index) const { assert(index < count); return data[index]; }
+    const T& Load(size_t index) const { assert(index < count); return data[index]; }  
+    void GetDimensions(uint32_t& outNumStructs, uint32_t& outStride) { outNumStructs = uint32_t(count); outStride = uint32_t(sizeof(T)); }
```

Comment:
I'm not 100% sure that `sizeof(T)` here for the stride is valid. It may be possible to create a structured buffer with, say, a 16-byte stride but then bind it for a `StructuredBuffer<float>` shader parameter. I'm not sure if that is allowed, but if it *is* then I'd assume that ends up working as a "view" to the first 4 bytes of each 16-byte element (so the element *size* is 4 bytes but the stride is 16).

I don't know for sure, though, and that certainly wouldn't be the 99% case, so this implementation is probably fine for now.